src/nepiada/utils/graphs.py

```diff
@@ -59,8 +74,68 @@ def update_graphs(self, agents):
             # Do not update the observation graph, because it has been configured to be static
             pass
 
-        # Update communication graph here
-        # Nothing to do, as we have defined communication graph to be static
+    def _update_comm_graph(self, agents):
+        """
+        If dynamic communication graph is enabled, this function will update the communication graph based on the current positions of the agents. 
+        It will also enforce a minimum number of agents that can communicate with each other, if the number of agents that can communicate with each other
+        is less than the minimum specified by the configuration.
+        If static communication graph, this function will do nothing since the communication graph is already configured to allow for full communication.
+    
+        params:
+            agents: A dictionary of agents, where the key is the agent's name and the value is the agent object.
+        """
+        if self.dynamic_comms:
+            # Reset the graph
+            self.comm = {agent: [] for agent in agents}
+            for agent_name, agent in agents.items():
+                x_coord = agent.p_pos[0]
+                y_coord = agent.p_pos[1]
+
+                distances_to_agents_not_added = []
+                # Update communication graph here
+                for other_agent_name, other_agent in agents.items():
+                    if other_agent_name != agent_name:
+                        # Calculate the distance between them
+                        delta_x = other_agent.p_pos[0] - x_coord
+                        delta_y = other_agent.p_pos[1] - y_coord
+
+                        distance = np.sqrt(delta_x**2 + delta_y**2)
+
+                        # If within communication radius add them to the graph
+                        # If not within the communication radius, add them to a list of agents that are not added,
+                        # for potential future addition based on enforcing minimum number of comm agents.
+                        if distance < self.dynamic_comms_radius:
+                            self.comm[agent_name].append(other_agent_name)
+                        else:
+                            distances_to_agents_not_added.append(
+                                (distance, other_agent_name)
+                            )
+                if len(self.comm[agent_name]) < self.dynamic_comms_enforce_minimum:
+                    # Sort the distances to agents not added in ascending order
+                    distances_to_agents_not_added.sort(key=lambda x: x[0])
```

Comment:
Observation: Assuming our enforced minimum is a small number the sorting might be a bottleneck here.

Reply:
In that edge case it may be, but this would be one of the better ways to do it. The flip side would be to store the non-added agents in a heap-like structure, but this would increase runtime in the average case when the minimums are already met. If the minimum is small, most of the time the minimum should already be met by the radius unless the radius is also small.

Reply:
We'll see how it affects our algorithm run times in the future. I am not sure if heap is the best way forward but we will cross that bridge when we come to it.

Reply:
Yeah I agree, technically the case where the sorting becomes a bottleneck shouldn't occur that much, and if it does occur, computation for the rest of NEPIADA should be less since it would be an environment with less agents and small radiuses.